Print unhandled exception in thread on stderr

dlech opened the pull request:
Regular unhandled exceptions are already printed on stderr. It seems
expected that unhandled exceptions in threads should also print to stderr.

dlech commented on py/modthread.c:
```diff
@@ -154,6 +154,8 @@ typedef struct _thread_entry_args_t {
     mp_obj_t args[];
 } thread_entry_args_t;
 
+extern const mp_print_t mp_stderr_print;
```
I'm guessing that you will not want the `extern` here.

I was not sure where to put it since `mp_stderr_print` is defined in `main.c`.

dlech commented:
Oops. I see now that this code is not just unix. I guess it is a bit more complicated than I thought.

pfalcon commented:
It's the same problem as https://github.com/micropython/micropython/issues/2838.

pfalcon commented:
(If it's a problem at all - no references made to the CPython behavior, so it's unclear what problem is being fixed here.)

dlech commented:
I'm working on developing a MicroPython library for LEGO MINDSTORMS EV3 as part of the http://www.ev3dev.org project. Parts of the platform we have developed expect error messages on stderr which gets redirected to a file or to the IDE. Otherwise, the message gets displayed on the screen, which is very small and hard to read, so you only see the last line of the exception. Or if the program crashes, you don't see anything at all.

CPython behavior is to print unhandled exceptions on threads to stderr.

dlech commented:
I have updated the pull request to use an existing option to enable this conditionally and also fix #2838.

I think the tests are failing because of issues with flushing stderr. If I run `exception_chain.py` manually, it produces the correct output.

```
$ ../ports/unix/micropython basics/exception_chain.py
Warning: exception chaining not supported
Caught Exception
$ cat basics/exception_chain.py.exp
Warning: exception chaining not supported
Caught Exception
```

dlech commented:
(hit the button too soon... continuing...)

```
$ cat exception_chain.py.out
Caught Exception
```

Either that or the tests are not capturing stderr. Although I see at least one place in `run-tests` where stderr is redirected to stdout.

dlech commented:
@pfalcon, it looks like I have all of the CI errors fixed, so this is ready for another look.

pfalcon commented:
> I'm working on developing a MicroPython library for LEGO MINDSTORMS EV3

Thanks, it's always helpful to know what people are working on and requests/patches come from.

> CPython behavior is to print unhandled exceptions on threads to stderr.

Makes sense to fix then, but should cover mentioned issue too, and not regress other ports.

pfalcon commented on py/modsys.c:
```diff
@@ -49,6 +49,7 @@ extern struct _mp_dummy_t mp_sys_stderr_obj;
 
 #if MICROPY_PY_IO && MICROPY_PY_SYS_STDFILES
 const mp_print_t mp_sys_stdout_print = {&mp_sys_stdout_obj, mp_stream_write_adaptor};
+const mp_print_t mp_sys_stderr_print = {&mp_sys_stderr_obj, mp_stream_write_adaptor};
```
I wouldn't do it like that. Why do we need this if we know that majority of ports don't have dedicated stderr stream and it defaults to same stream as stdout/plat? This would make a difference if we allowed overriding sys.stderr, but we don't, and it will take its own time and usecases before we will. Also, what about the way the unix port prints exceptions currently? There's one mp_stderr_print, you introduce another => mess.

I'd suggest to start low-profile, like we usually do in uPy: let MP_PYTHON_ERROR_PRINTER default to &mp_plat_print, like before. But a port can define it to something else, which the unix port will do.

pfalcon commented:
> so this is ready for another look.

Yes, please have a look at https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md regarding commit message formatting ;-). Also, please squash commits together.

dlech commented:
Thank you for the review. Should I move `mp_stderr_print` out of `ports/unix/main.c`? If so, where? Or, if not, which header file should I add the definition to?

dlech committed:
py: add config option to print warnings/errors to stderr

This adds a new configuration option to print runtime warnings and errors to
stderr. On Unix, CPython prints warnings and unhandled exceptions to stderr,
so the unix port here is configured to use this option.

The unix port already printed unhandled exceptions on the main thread to
stderr. This patch fixes unhandled exceptions on other threads and warnings
(issue #2838) not printing on stderr.

Additionally, a couple tests needed to be fixed to handle this new behavior.
This is done by also capturing stderr when running tests.

pfalcon commented:
> Should I move mp_stderr_print out of ports/unix/main.c?

Usually, fix which makes minimal changes is the right one. Not always - but then there always should be arguments why "bigger" change is "better".

> which header file should I add the definition to?

mpconfigport.h, see the existing stuff there for an idea how it can be done.

dlech commented:
I have updated the PR to include a new `MICROPY_ERROR_PRINTER` configuration option.

dpgeorge commented:
Thanks!  Merged in 62849b7010abffb7b0a9c9875930efe7cb77519c
